### PR TITLE
bpo-35537: use os.posix_spawn in subprocess

### PR DESCRIPTION
**Pull request by nanjekyejoannah**
This patch uses os.posix_spawn in subprocess.

posix_spawn() is faster but it's also safer: it allows us to do a lot of "actions" before exec(), before executing the new program. For example, you can close files and control signals. 

@vstinner this is the patch without any options.

https://bugs.python.org/issue35537

### CODE REVIEW

**Comment by vstinner** [comment]
@nanjekyejoannah: Thanks! Can you please open an issue on bugs.python.org to get a bug number? Then add "bpo-xxx: " to the title of your PR.

Can you also please add a NEWS entry? You can use the "blurb" tool: https://devguide.python.org/committing/#what-s-new-and-news-entries

File: Lib/subprocess.py
Comment (vstinner) [inline Lib/subprocess.py]: "and not restore_signals" would be better. I prefer to avoid "is" to compare with True/False.

File: Lib/subprocess.py
Comment (vstinner) [inline Lib/subprocess.py]: Is it really needeed to copy os.environ?

File: Misc/NEWS.d/next/Library/2018-12-19-17-20-03.bpo-35537.z4E7aA.rst
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2018-12-19-17-20-03.bpo-35537.z4E7aA.rst]: For better rendering (create links), I suggest:

```suggestion
:class:`subprocess.Popen` can now use :func:`os.posix_spawn` in some cases.
```

**Review by vstinner — APPROVED** [review]
LGTM, but I will wait until minor remarks are addressed.

File: Lib/subprocess.py
Comment (izbyshev) [inline Lib/subprocess.py:1432]: The condition is too weak. It doesn't check for redirections at all, so e.g. `stdout=subprocess.PIPE` won't work: it will leave both pipe ends open in the parent, but the child won't inherit anything because both ends have `O_CLOEXEC`.

Another problem is that `os.posix_spawn` never looks in `PATH`, so e.g. `subprocess.call(['sleep', '10'], close_fds=False, restore_signals=False)` doesn't work with this patch. 

Looks like some tests that would catch the above are needed :)

Comment (vstinner) [reply Lib/subprocess.py:1432]: Right. I suggest to add more tests:

```
and os.path.isabs(executable)
and self.stdin is self.stdout is self.stderr is None)
```

File: Lib/subprocess.py
Comment (vstinner) [inline Lib/subprocess.py]: I suggest to pass (executable, args) to follow os.posix_spawn() API. The function has not return value, so I suggest to write instead:

```suggestion
                 self._posix_spawn(executable, args)
                 return
```

File: Lib/subprocess.py
Comment (vstinner) [reply Lib/subprocess.py]: I suggest to pass directly os.environ to os.posix_spawn().

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/subprocess.py
Comment (vstinner) [reply Lib/subprocess.py:1432]: > Looks like some tests that would catch the above are needed :)

test_subprocess is quite complete. But with this PR, posix_spawn() is only used under very specific conditions: restore_signals=False and close_fds=False. This case is rare.

@nanjekyejoannah is my mentoree and we are working together to slowly use posix_spawn() in more and more cases. I prefer to work step by step, and so support more and more options, but start from the bare minimum, and check at each step that everything is fine.

Comment (izbyshev) [reply Lib/subprocess.py:1432]: `os.path.isabs(executable)` is a bit strong. Using `os.sep not in executable` is enough: it's exactly what `execvp` and `posix_spawnp` do when they decide whether to do `PATH` lookup. But IMO even better would be to add `os.posix_spawnp` and use it instead.

> test_subprocess is quite complete. But with this PR, posix_spawn() is only used under very specific conditions: restore_signals=False and close_fds=False. This case is rare.

It's a bit scary for me that basic functionality broke (albeit under these rare conditions) and no test failed. Basically, it indicates that if users pass `restore_signals=False`, they're on their own (`close_fds` seems to be better tested).

Comment (vstinner) [reply Lib/subprocess.py:1432]: Don't worry, os.posix_spawn() support will be much more complete before Python 3.8 final. I'm working with @nanjekyejoannah to support restore_signals=True for example ;-)

Comment (vstinner) [reply Lib/subprocess.py:1432]: > But IMO even better would be to add os.posix_spawnp and use it instead.

Aha, I didn't know posix_spawnp(), that's interesting. It is available on macOS and FreeBSD.

Again, I prefer to keep the first implementation as simple as possible. os.posix_spawnp() can be added later.

> os.path.isabs(executable) is a bit strong. Using os.sep not in executable is enough: it's exactly what execvp and posix_spawnp do when they decide whether to do PATH lookup.

shutil.which() uses os.path.dirname(). So I suggest to check "and os.path.dirname(executable)" instead of "and os.path.isabs(executable)".

Comment (vstinner) [reply Lib/subprocess.py:1432]: Update, according to the discussion, the following check would be better:

```
and os.path.dirname(executable)
and self.stdin is self.stdout is self.stderr is None)
```

**Review by vstinner — COMMENTED** [review]
Can you please add a sentence to the Optimization section of Doc/whatsnew/3.8.rst? Something like:

"subprocess.Popen is now up to 60x faster on Linux and 106x faster on FreeBSD when os.posix_spawn() can be used."

**Comment by serhiy-storchaka** [comment]
60x and 106x are from particular examples. If you use different examples (allocate 1 GiB or 4 GiB of memory in the parent) you can get different numbers. Concrete numbers depend on a program and do not have meaning without referring to concrete program.

**Comment by vstinner** [comment]
> 60x and 106x are from particular examples. If you use different examples (allocate 1 GiB or 4 GiB of memory in the parent) you can get different numbers. Concrete numbers depend on a program and do not have meaning without referring to concrete program.

It's hard to measure the average speedup, but IMHO it's nice to say that it can be way faster in some cases. So do you want to explain it differently? For example:

"subprocess.Popen is now up to 60x faster on Linux and 106x faster on FreeBSD when os.posix_spawn() can be used, the parent has large memory footprint (ex: 2 GiB), and the child process execution time is short."

**Comment by serhiy-storchaka** [comment]
I suggest to not use any precise numbers: "subprocess.Popen is now significantly faster" or "subprocess.Popen can be now tens times faster".

**Comment by nanjekyejoannah** [comment]
@vstinner I will reopen did it by mistake.

**Comment by vstinner** [comment]
@nanjekyejoannah: this PR is now empty, it doesn't contain any commit :-( It seems like your removed your commit. You can use "git reflog" or open  .git/logs/HEAD to recover your removed commit.

File: Lib/subprocess.py
Comment (vstinner) [inline Lib/subprocess.py]: nitpick: this variable is useless, just pass directly os.environ to posix_spawn().

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
I will merge this PR tomorrow except if someone sees a good reason to not merge it: https://bugs.python.org/issue35537#msg332239

File: Lib/subprocess.py
Comment (izbyshev) [inline Lib/subprocess.py:1433]: Checking `self.std*` is still not enough. Those fields are for the parent's ends of pipes, but if a caller redirects to a descriptor, there is no such end. For example, `subprocess.call(['/bin/sleep', '1000'], stdout=fd, close_fds=False, restore_signals=False)` doesn't work properly.

Comment (vstinner) [reply Lib/subprocess.py:1433]: @izbyshev: Oh right! I fixed the test:  test  (p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite) instead of (self.stdin, self.stdout, self.stderr). Is it better now?

**Comment by vstinner** [comment]
@nanjekyejoannah: I pushed directly a fix into your branch to fix the latest reported bug about standard streams ;-)

**Comment by vstinner** [comment]
> I will merge this PR tomorrow except if someone sees a good reason to not merge it: https://bugs.python.org/issue35537#msg332239

Hum, @izbyshev found another issue (already now fixed) with this PR. Since I will be offline next 2 weeks, I prefer to not merge yet since I would be unable to handle any issue caused by the change. I now prefer to wait until I'm back from holiday to merge the PR.

In the meanwhile, if another core dev feels brave enough to merge it, please go ahead ;-)

**Comment by nanjekyejoannah** [comment]
okay noted.

File: Lib/subprocess.py
Comment (izbyshev) [reply Lib/subprocess.py:1433]: Yes, LGTM now.

There is another potential problem with `posix_spawn`: [asynchronous error notification](http://pubs.opengroup.org/onlinepubs/9699919799/functions/posix_spawn.html). `_posixsubprocess` uses a pipe to notify the parent about all errors up to and including `execve` errors, so its style is synchronous. Glibc and musl use the pipe-based notification too, but they are not required to do that, and I don't know what other implementations do.

**Comment by vstinner** [comment]
I expect that posix_spawn() sets errno as does execv(). I tested a non
existent executable and I got the expected exception and traceback.

The pipe matters more when preexec_fn is used to return the Python
traceback of the child and the exception type. But preeexec_fn doesn't make
sense with (cannot be supported by) posix_spawn().

-- 
Night gathers, and now my watch begins. It shall not end until my death.

**Comment by izbyshev** [comment]
> I expect that posix_spawn() sets errno as does execv()

As I explained, this behavior is not required by POSIX, and the opposite behavior is allowed.  Recent glibc- and musl-based Linux behave as you expect, but for glibc, this was added in [2017](https://sourceware.org/git/?p=glibc.git;a=commit;h=ccfb2964726512f6669fea99a43afa714e2e6a80). In my test on Ubuntu 14.04, `./python -c "import subprocess; subprocess.call(['/xxx'], close_fds=False, restore_signals=False)"` silently returns with zero exit code.

FreeBSD also [seems not to use](https://github.com/freebsd/freebsd/blob/98e988b9a4570300cb42b11e11564940d6af920b/lib/libc/gen/posix_spawn.c#L225) any synchronous notification.

Other systems where `posix_spawn` is a syscall (macOS?) probably behave as you expect, but I've not checked.

**Comment by vstinner** [comment]
I merged PR #11452 instead.

**Pull request closed without merging**
